authorize_helper.go

```diff
@@ -41,11 +41,17 @@ import (
 func GetRedirectURIFromRequestValues(values url.Values) (string, error) {
 	// rfc6749 3.1.   Authorization Endpoint
 	// The endpoint URI MAY include an "application/x-www-form-urlencoded" formatted (per Appendix B) query component
-	redirectURI, err := url.QueryUnescape(values.Get("redirect_uri"))
+	redirectURI, err := url.Parse(values.Get("redirect_uri"))
 	if err != nil {
 		return "", errors.WithStack(ErrInvalidRequest.WithHint(`The "redirect_uri" parameter is malformed or missing.`).WithCause(err).WithDebug(err.Error()))
 	}
-	return redirectURI, nil
+	rawQuery, err := url.QueryUnescape(redirectURI.RawQuery)
```

Comment:
But I do not think this is necessary. `values` should already include correctly parsed parameter from the form payload, so no further unescaping of the `redirect_uri` should be necessary. We should just pass it on as-is. We should just validate that the `redirect_uri` is valid URL.

Do you have a test case where `QueryUnescape` is necessary?

Reply:
Without escaping the query part, one [existing test](https://github.com/ory/fosite/blob/9cecb109a338da8a5c1bb16a10bb00344dba89ba/authorize_helper_test.go#L56) case fails. 
According to this test case, During the client registration, if redirct_uri has an url-encoded query parameter, it is expected to be escaped. I am not sure if it is expected according to the spec.

Reply:
But that unescaping should be done by accessing the query string using `Query()`, no? So form payload unparses it once as a whole `redirect_uri` parameter. And then `URL` as result of `url.Parse` has `Query()` method which unescapes when needed. There is no point in doing `redirectURI.RawQuery = url.QueryUnescape(redirectURI.RawQuery)`.

I think the confusion here is from this sentence in the spec:

> The endpoint URI MAY include an "application/x-www-form-urlencoded" formatted (per Appendix B) query component ([RFC3986] Section 3.4), which MUST be retained when adding additional query parameters.

I think this does not mean that you do `urlescape("param1=value1&param2=value2")` but that it has to be `"param1=" + urlescape(value1) + "&param2=" + urlescape(value2)`. So the word there is `formatted`, not `escaped`. But in fact, both of those forms are exactly equivalent and should be processed the same by `url.Query()` call. See [this seciton](https://tools.ietf.org/html/rfc3986#section-3.4) (referenced from [RFC section](https://tools.ietf.org/html/rfc6749#section-3.1.2) referenced from the test you linked above) which says:

> However, as query components are often used to carry identifying information in the form of "key=value" pairs and one frequently used value is a reference to another URI, it is sometimes better for usability to avoid percent-encoding those characters.

So urlescaping also `+` and `=` characters in the query string is fine. But it is not necessary for usability. But both of those are demeed to be `"application/x-www-form-urlencoded" formatted`. In any case, we do not have to double unescape things. We just have to make sure that we append additional query values to existing. Which we should be doing by calling `url.Query()` (which would unescape things), add additional ones, and then encode them back. Which I think we are doing, no?

So, no need for double decoding here.

Reply:
Yes. I completely agree with you. So, we can get rid of `GetRedirectURIFromRequestValues` method and the associated test case.

Reply:
No, we can keep it, just to verify the redirect URI.

You should probably update the test case accordingly.